[#6193] Delay server only logs warning when byte limit is not defined in server config (main)

korydraughn opened the pull request.

alanking reviewed (dismissed):
\# it up

korydraughn commented:
cool cool.

@trel do we still want to add the default to the server config?

trel commented:
I think yes, because it provides bumpers and reduces drift over time as compared to the shipped template...  (which suggests we should do this with all new introduced configurations).

But... @kript says it affects his configuration-management scripts and deployments...   (I think that's a signal that the configuration management scripts should be updated itself...)

So, yes?

korydraughn commented:
Will make it happen.

korydraughn commented:
This PR should be merged after #6274.

korydraughn commented on scripts/irods/upgrade_configuration.py:
```diff
@@ -92,70 +102,60 @@ def upgrade(irods_config):
-        # Load server_config.json.
-        with open(paths.server_config_path()) as f:
-            server_config = json.load(f)
+    if old_version_tuple[:2] <= (4, 3):
+        # Build a new server_config.json file using server_config.json.template as a base.
+        # Overwrite all configuration properties in the template with the properties from
+        # server_config.json.
+        server_config = load_server_config_with_explicit_defaults()
```
How do we feel about this logic?

It attempts to always provide a server config that contains all options explicitly defined. This code path is always invoked when dealing with a 4.3+ server. Devs are expected to update `server_config` from here on.

A better solution is probably to go on and write `server_config` to disk. Then the code is free to reload the config and make changes as needed.

trel replied:
Interesting - so the approach is to populate the template, rather than 'sniff and inject' into the existing file content.

I'm most concerned about missing things that we don't know about... like environment variables that someone might be depending on.

Building on what they already have in place is a safer starting point, I think...

korydraughn replied:
Are you talking about environment variables defined in server_config.json?

If so, this should work. We start with the template and overwrite every matching key in the local server config.

All options the admin set originally should be maintained.

korydraughn replied:
I've adjusted the logic so that it is easier to maintain and understand.

korydraughn commented on scripts/irods/upgrade_configuration.py:
```diff
@@ -92,70 +102,60 @@ def upgrade(irods_config):
+        if write_changes_to_disk:
             irods_config.commit(server_config, paths.server_config_path(), make_backup=True)
 
+        # TODO Should we rename hosts_config.json and host_access_control_config.json so that
+        # the upgrade process does not process them anymore?
```
Thoughts?

trel replied:
I think upgrade should sniff if those stanzas/values already exist in server_config.json and do the conversion if not found.  I don't think we should rename or delete these other two 'legacy' files?

trel replied:
Renaming them could be elegant... perhaps appending ... `.absorbed_by_4.3.0`?   Hmm....

But definitely, they'd be ignored.

korydraughn replied:
Thinking about it more, how about `.absorbed_by_4.3.0_server_config`?

korydraughn replied:
I've adjusted the code to append `.absorbed_by_4.3.0_server_config` instead of the former because it is more descriptive about what happened.

korydraughn commented:
Running core tests ...

korydraughn commented:
All core tests passed.

Will add the #'s after #6274 is merged.

korydraughn commented:
Using a docker container, I've verified that this PR handles the following during an upgrade from 4.2.11 to 4.3.0:
- Renamed properties are mapped to the value of the old properties
- All missing properties are added with proper defaults
- Properties that aren't used by the server anymore are removed
- Unknown properties are carried forward as-is
- host_access_control_config.json is absorbed into server_config.json and renamed
- hosts_config.json is absorbed into server_config.json and renamed
- A backup copy of server_config.json is generated before it's overwritten
---
4.2.11 VERSION.json file:
```json
{
    "catalog_schema_version": 8,
    "commit_id": "bc6f9f1cdef6c4ec01ea14402428988892615321",
    "configuration_schema_version": 3,
    "installation_time": "2022-04-03T12:40:22.787456",
    "irods_version": "4.2.11"
}
```

4.2.11 server config before upgrade:
- Contains unknown properties (simple and nested)
- Contains modified values (not an exhaustive list)
    - `maximum_number_of_concurrent_rule_engine_server_processes`
    - `rule_engine_server_sleep_time_in_seconds`
    - `environment_variables`
- Some properties are missing (not an exhaustive list)
    - `dns_cache`
    - `hostname_cache`
    - `client_api_whitelist_policy`
- hosts_config.json is an external file
- host_access_control_config.json is an external file
```json
{
    "UNKNOWN_OPTION_0": [4, 3, 2, 1, 1, 2, 3, 4],
    "UNKNOWN_OPTION_1": [
        {
            "key_1": "value_1",
            "key_2": "value_2",
            "key_3": "value_3"
        },
        {
            "value_9": "key_9",
            "value_2": "key_2",
            "value_7": "key_7",
            "other": {
                "stuff": "this is important"
            }
        }
    ],
    "advanced_settings": {
        "default_log_rotation_in_days": 5, 
        "default_number_of_transfer_threads": 4, 
        "default_temporary_password_lifetime_in_seconds": 120, 
        "maximum_number_of_concurrent_rule_engine_server_processes": 8, 
        "maximum_size_for_single_buffer_in_megabytes": 32, 
        "maximum_temporary_password_lifetime_in_seconds": 1000, 
        "rule_engine_server_execution_time_in_seconds": 120, 
        "rule_engine_server_sleep_time_in_seconds": 90, 
        "transfer_buffer_size_for_parallel_transfer_in_megabytes": 4, 
        "transfer_chunk_size_for_parallel_transfer_in_megabytes": 40
    }, 
    "catalog_provider_hosts": [
        "3cd3c95fbf79"
    ], 
    "catalog_service_role": "provider", 
    "default_dir_mode": "0750", 
    "default_file_mode": "0600", 
    "default_hash_scheme": "SHA256", 
    "default_resource_name": "demoResc", 
    "environment_variables": {
        "SOME_ENV_VAR_1": "SOME_ENV_VAR_VALUE_1",
        "SOME_ENV_VAR_2": "SOME_ENV_VAR_VALUE_2"
    }, 
    "federation": [], 
    "match_hash_policy": "compatible", 
    "negotiation_key": "32_byte_server_negotiation_key__", 
    "plugin_configuration": {
        "authentication": {}, 
        "database": {
            "postgres": {
                "db_host": "catalog", 
                "db_name": "ICAT", 
                "db_odbc_driver": "PostgreSQL ANSI", 
                "db_password": "testpassword", 
                "db_port": 5432, 
                "db_username": "irods"
            }
        }, 
        "network": {}, 
        "resource": {}, 
        "rule_engines": [
            {
                "instance_name": "irods_rule_engine_plugin-irods_rule_language-instance", 
                "plugin_name": "irods_rule_engine_plugin-irods_rule_language", 
                "plugin_specific_configuration": {
                    "re_data_variable_mapping_set": [
                        "core"
                    ], 
                    "re_function_name_mapping_set": [
                        "core"
                    ], 
                    "re_rulebase_set": [
                        "core"
                    ], 
                    "regexes_for_supported_peps": [
                        "ac[^ ]*", 
                        "msi[^ ]*", 
                        "[^ ]*pep_[^ ]*_(pre|post|except|finally)"
                    ]
                }, 
                "shared_memory_instance": "irods_rule_language_rule_engine"
            }, 
            {
                "instance_name": "irods_rule_engine_plugin-cpp_default_policy-instance", 
                "plugin_name": "irods_rule_engine_plugin-cpp_default_policy", 
                "plugin_specific_configuration": {}
            }
        ]
    }, 
    "rule_engine_namespaces": [
        ""
    ], 
    "schema_name": "server_config", 
    "schema_validation_base_uri": "file:///var/lib/irods/configuration_schemas", 
    "schema_version": "v3", 
    "server_control_plane_encryption_algorithm": "AES-256-CBC", 
    "server_control_plane_encryption_num_hash_rounds": 16, 
    "server_control_plane_key": "32_byte_server_control_plane_key", 
    "server_control_plane_port": 1248, 
    "server_control_plane_timeout_milliseconds": 10000, 
    "server_port_range_end": 20199, 
    "server_port_range_start": 20000, 
    "xmsg_port": 1279, 
    "zone_auth_scheme": "native", 
    "zone_key": "TEMPORARY_ZONE_KEY", 
    "zone_name": "tempZone", 
    "zone_port": 1247, 
    "zone_user": "rods"
}
```

server_config.json after upgrade to 4.3.0:
```json
{
    "UNKNOWN_OPTION_0": [
        4,
        3,
        2,
        1,
        1,
        2,
        3,
        4
    ],
    "UNKNOWN_OPTION_1": [
        {
            "key_1": "value_1",
            "key_2": "value_2",
            "key_3": "value_3"
        },
        {
            "other": {
                "stuff": "this is important"
            },
            "value_2": "key_2",
            "value_7": "key_7",
            "value_9": "key_9"
        }
    ],
    "advanced_settings": {
        "default_log_rotation_in_days": 5,
        "default_number_of_transfer_threads": 4,
        "default_temporary_password_lifetime_in_seconds": 120,
        "delay_rule_executors": [],
        "delay_server_sleep_time_in_seconds": 90,
        "dns_cache": {
            "eviction_age_in_seconds": 3600,
            "shared_memory_size_in_bytes": 5000000
        },
        "hostname_cache": {
            "eviction_age_in_seconds": 3600,
            "shared_memory_size_in_bytes": 2500000
        },
        "maximum_size_for_single_buffer_in_megabytes": 32,
        "maximum_size_of_delay_queue_in_bytes": 0,
        "maximum_temporary_password_lifetime_in_seconds": 1000,
        "number_of_concurrent_delay_rule_executors": 8,
        "stacktrace_file_processor_sleep_time_in_seconds": 10,
        "transfer_buffer_size_for_parallel_transfer_in_megabytes": 4,
        "transfer_chunk_size_for_parallel_transfer_in_megabytes": 40
    },
    "catalog_provider_hosts": [
        "3cd3c95fbf79"
    ],
    "catalog_service_role": "provider",
    "client_api_whitelist_policy": "enforce",
    "default_dir_mode": "0750",
    "default_file_mode": "0600",
    "default_hash_scheme": "SHA256",
    "default_resource_name": "demoResc",
    "environment_variables": {
        "SOME_ENV_VAR_1": "SOME_ENV_VAR_VALUE_1",
        "SOME_ENV_VAR_2": "SOME_ENV_VAR_VALUE_2"
    },
    "federation": [],
    "host_access_control": {
        "access_entries": []
    },
    "host_resolution": {
        "host_entries": []
    },
    "log_level": {
        "agent": "info",
        "agent_factory": "info",
        "api": "info",
        "authentication": "info",
        "database": "info",
        "delay_server": "info",
        "legacy": "info",
        "microservice": "info",
        "network": "info",
        "resource": "info",
        "rule_engine": "info",
        "server": "info"
    },
    "match_hash_policy": "compatible",
    "negotiation_key": "32_byte_server_negotiation_key__",
    "plugin_configuration": {
        "authentication": {},
        "database": {
            "postgres": {
                "db_host": "catalog",
                "db_name": "ICAT",
                "db_odbc_driver": "PostgreSQL ANSI",
                "db_password": "testpassword",
                "db_port": 5432,
                "db_username": "irods"
            }
        },
        "network": {},
        "resource": {},
        "rule_engines": [
            {
                "instance_name": "irods_rule_engine_plugin-irods_rule_language-instance",
                "plugin_name": "irods_rule_engine_plugin-irods_rule_language",
                "plugin_specific_configuration": {
                    "re_data_variable_mapping_set": [
                        "core"
                    ],
                    "re_function_name_mapping_set": [
                        "core"
                    ],
                    "re_rulebase_set": [
                        "core"
                    ],
                    "regexes_for_supported_peps": [
                        "ac[^ ]*",
                        "msi[^ ]*",
                        "[^ ]*pep_[^ ]*_(pre|post|except|finally)"
                    ]
                },
                "shared_memory_instance": "irods_rule_language_rule_engine"
            },
            {
                "instance_name": "irods_rule_engine_plugin-cpp_default_policy-instance",
                "plugin_name": "irods_rule_engine_plugin-cpp_default_policy",
                "plugin_specific_configuration": {}
            }
        ]
    },
    "rule_engine_namespaces": [
        ""
    ],
    "schema_name": "server_config",
    "schema_validation_base_uri": "file:///var/lib/irods/configuration_schemas",
    "schema_version": "v4",
    "server_control_plane_encryption_algorithm": "AES-256-CBC",
    "server_control_plane_encryption_num_hash_rounds": 16,
    "server_control_plane_key": "32_byte_server_control_plane_key",
    "server_control_plane_port": 1248,
    "server_control_plane_timeout_milliseconds": 10000,
    "server_port_range_end": 20199,
    "server_port_range_start": 20000,
    "xmsg_port": 1279,
    "zone_auth_scheme": "native",
    "zone_key": "TEMPORARY_ZONE_KEY",
    "zone_name": "tempZone",
    "zone_port": 1247,
    "zone_user": "rods"
}
```
The contents of /etc/irods after upgrade:
```bash
$ ls -l /etc/irods/                  
total 76
-rw-r--r-- 1 irods irods  5026 Apr  3 12:40 core.dvm
-rw-r--r-- 1 irods irods   831 Apr  3 12:40 core.fnm
-rw-r--r-- 1 irods irods 38846 Apr  3 12:40 core.re
-rw-r--r-- 1 irods irods   106 Apr  3 12:40 host_access_control_config.json.absorbed_by_4.3.0_server_config
-rw-r--r-- 1 irods irods    90 Apr  3 12:40 hosts_config.json.absorbed_by_4.3.0_server_config
-rw------- 1 irods irods  4792 Apr  3 13:06 server_config.json
-rw-rw-r-- 1 irods irods  3894 Apr  3 13:06 server_config.json.prev.1648991186.8834982
-rw-r--r-- 1 irods irods    64 Apr  3 12:40 service_account.config
```

korydraughn commented:
One thing that needs to be considered is how we handle the `client_api_whitelist_policy` option.

The logic does not seem correct according to the docs.
https://github.com/irods/irods/blob/eb361f5e49e817ff498dbb417eb935030e5b2322/server/core/src/client_api_whitelist.cpp#L188-L206
See what is mentioned in the docs about this property here: https://docs.irods.org/4.2.11/system_overview/configuration/

The docs say the default is to **enforce** the allowlist. However, the code says... if the property does not exist, ignore the allowlist entirely.

I'll create an issue for this and we can handle it in a separate PR.

korydraughn commented:
Issue created: #6295

trel commented:
What does the suffix represent here?
```
server_config.json.prev.1648991186.8834982
```

trel commented:
consider getting whitelist, itself, renamed as part of this PR for #5926 as well.

korydraughn commented:
> What does the suffix represent here?
> 
> ```
> server_config.json.prev.1648991186.8834982
> ```

It is the epoch as a floating point number. The last element might be nanoseconds.

From Python 3 docs: https://docs.python.org/3/library/time.html#time.time

See lines 295-296.
https://github.com/irods/irods/blob/eb361f5e49e817ff498dbb417eb935030e5b2322/scripts/irods/configuration.py#L287-L299

korydraughn commented:
After looking at the code, I think we can safely truncate the microseconds from the epoch suffix. The files are only encountered once. This means we aren't at risk of losing the original config file.

korydraughn commented:
With the most recent commit, here is the directory listing without the level of time resolution on the file backups.
```bash
$ ls -l /etc/irods
total 76
-rw-r--r-- 1 irods irods  5026 Apr  4 22:14 core.dvm
-rw-r--r-- 1 irods irods   831 Apr  4 22:14 core.fnm
-rw-r--r-- 1 irods irods 38846 Apr  4 22:14 core.re
-rw-r--r-- 1 irods irods   106 Apr  4 22:14 host_access_control_config.json.absorbed_by_4.3.0_server_config
-rw-r--r-- 1 irods irods    90 Apr  4 22:14 hosts_config.json.absorbed_by_4.3.0_server_config
-rw------- 1 irods irods  4354 Apr  4 22:22 server_config.json
-rw-rw-r-- 1 irods irods  3881 Apr  4 22:22 server_config.json.prev.1649110934
-rw-r--r-- 1 irods irods    64 Apr  4 22:14 service_account.config
```

trel commented:
but... this is more human-friendly?

```python
>>> datetime.datetime.utcnow().strftime("%Y%m%dT%H%M%SZ")
'20220404T232420Z'
```

korydraughn commented:
True.

I ended up with the epoch seconds scheme only because we have other things that use epoch seconds (and milliseconds too) as a filename suffix. For example, the async-signal-safe stacktrace file pickup implementation.

Should all filename suffixes based on time use the same convention or is it ok for them to be different?

korydraughn commented:
While I'd like for things to use the same scheme, I'm starting to lean towards it being ok for the system to use a variety of them. Different tasks require different solutions.

Thoughts?

trel commented:
in general, my heuristic is that if it's for human consumption, it provides a little more info via inspection to have iso8601.

epoch is for the computer - and languages can all handle it without ambiguity.

and subseconds if it matters for collision avoidance.

korydraughn commented:
Will adjust the code to use the human readable form.

trel commented:
consider https://github.com/irods/irods/issues/6298 as well, please - this feels like the right moment for it.

korydraughn commented on configuration_schemas/v4/server_config.json.in:
```diff
@@ -3,15 +3,50 @@
             "uniqueItems": true
         },
         "catalog_service_role": {"enum": ["consumer","provider"]},
+        "client_api_allowlist_policy": {"enum": ["disabled", "enforce"]},
```
Q. Any thoughts around this being treated as an enumeration?
Q. How do we feel about the word "disabled"?

Before this PR, we only treated this property as a string. Its value only mattered when it was set to "enforce". Other values were equivalent to "not enforce".

trel replied:
I think this is fine.   It limits the choices, and is clear.

Only downside to this... if we come up with another 'mode', we'll have to bump the schema version.   But I cannot think of what that would be at the moment.

korydraughn replied:
I can revert it to a string. That provides more flexibility if we feel we may need it.

trel replied:
it's basically a boolean for whether we want to make the thing 'go'.   So, I like 'disabled' and 'enforce'.  Let's do it.

korydraughn replied:
Cool cool.

korydraughn commented:
All core tests pass.

Once we have an answer for the `client_api_allowlist_policy` property, I'll update and add the #'s.

korydraughn commented:
Added #'s

Woohoo!
